# EVALUATION THE VAR TIME SERIES MODEL

# import data

import pandas as pd
import numpy 
import statsmodels
import matplotlib.pyplot as plt


df = pd.read_csv('West_Africa_temperature_1980_2022.csv')

In [12]:
#Transpose the data
df = df.T

# Set the second row as the header
new_header = df.iloc[0]  # Grab the first row for the header

df = df[1:]  # Take the data less the header row

df.columns = new_header  #Set the header row as the df header

df_main = df[4:]

#The temperatures in the dataframe are all abjects.
#Lets convert the objects to float
df_main = df_main.astype(float)

In [13]:
df_main.head()

STATION              UV000005502  UV000005507  UV000005522  UV000065501  \
1980-01-01 00:00:00        23.89        24.44        26.67        21.67   
1980-01-02 00:00:00        24.44        24.44        27.22        24.44   
1980-01-03 00:00:00        24.44        25.00        26.11        22.78   
1980-01-04 00:00:00        23.33        25.00        26.11        23.33   
1980-01-05 00:00:00        23.89        25.56        27.22        23.89   

STATION              UV000065516  UVM00065503  UVM00065510  UVM00065518  \
1980-01-01 00:00:00        24.44        23.33        24.44        26.67   
1980-01-02 00:00:00        26.11        25.00        26.11        26.67   
1980-01-03 00:00:00        25.00        25.56        25.00        26.67   
1980-01-04 00:00:00        24.44        24.44        25.00        26.67   
1980-01-05 00:00:00        25.56        25.56        26.67        26.67   

STATION              IV000005557  IV000005562  ...  SGM00061679  SGM00061695  \
1980-01-01 00:00:00        25.56        24.44  ...        26.67        23.33   
1980-01-02 00:00:00        25.56        27.22  ...        26.67        24.44   
1980-01-03 00:00:00        26.11        27.22  ...        26.67        23.33   
1980-01-04 00:00:00        26.67        26.67  ...        26.67        21.67   
1980-01-05 00:00:00        26.11        26.67  ...        24.44        24.44   

STATION              SGM00061697  SGM00061698  SGM00061699  SL000061856  \
1980-01-01 00:00:00        26.67        28.89        30.00        26.67   
1980-01-02 00:00:00        26.67        27.78        26.11        26.67   
1980-01-03 00:00:00        26.67        28.89        27.78        26.67   
1980-01-04 00:00:00        26.67        23.33        28.33        26.67   
1980-01-05 00:00:00        26.67        23.33        27.22        26.67   

STATION              LIM00065660  GHM00065472  CVM00008594  GV000001832  
1980-01-01 00:00:00        27.78        26.11        21.67        26.11  
1980-01-02 00:00:00        27.78        28.89        22.22        26.11  
1980-01-03 00:00:00        27.78        28.89        21.11        26.11  
1980-01-04 00:00:00        27.78        28.89        21.67        26.11  
1980-01-05 00:00:00        27.78        28.89        21.67        26.11  

[5 rows x 72 columns]

Since we have already established that the data above is stationery in the previous notebook, and we have decided to use a lag of 4, we can go ahead and slit the data into train and test set so that we can evaluate the model.

In [14]:
#Split the data

train_ratio = 0.8
split_index = int(len(df_main) * train_ratio)
train = df_main[:split_index]
test = df_main[split_index:]

In [18]:
# Initialize and fit the VAR model
from statsmodels.tsa.api import VAR

model = VAR(train)
results = model.fit(maxlags=4, ic='aic')
# maxlags is the maximum lag order, ic='aic' tells the model to use Akaike Information Criterion to select the best lag value

C:\Users\USER\anaconda3\Lib\site-packages\statsmodels\tsa\base\tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency D will be used.
  self._init_dates(dates, freq)


In [19]:
#Make prediction

predictions = results.forecast(train.values[-results.k_ar:], steps=len(test))

In [25]:
#Convert prediction to a dataframe

pred_df = pd.DataFrame(predictions, index=test.index, columns=test.columns)

In [27]:
pred_df

STATION              UV000005502  UV000005507  UV000005522  UV000065501  \
2014-05-26 00:00:00    31.731527    29.883788    28.327494    32.616965   
2014-05-27 00:00:00    33.091115    30.857352    28.878145    34.480544   
2014-05-28 00:00:00    33.260045    30.867647    28.732124    34.642013   
2014-05-29 00:00:00    33.039837    30.694143    28.486084    34.454695   
2014-05-30 00:00:00    33.116031    30.871830    28.629986    34.523976   
...                          ...          ...          ...          ...   
2022-12-27 00:00:00    29.186504    28.489782    27.640972    29.532933   
2022-12-28 00:00:00    29.186504    28.489782    27.640972    29.532933   
2022-12-29 00:00:00    29.186504    28.489782    27.640972    29.532933   
2022-12-30 00:00:00    29.186504    28.489782    27.640972    29.532933   
2022-12-31 00:00:00    29.186504    28.489782    27.640972    29.532933   

STATION              UV000065516  UVM00065503  UVM00065510  UVM00065518  \
2014-05-26 00:00:00    30.139896    30.807145    29.261668    30.043050   
2014-05-27 00:00:00    30.963134    32.107618    29.838706    30.487165   
2014-05-28 00:00:00    30.850014    32.055806    29.493107    30.280754   
2014-05-29 00:00:00    30.870178    31.577454    29.311420    30.042435   
2014-05-30 00:00:00    30.940639    31.757012    29.245702    30.033092   
...                          ...          ...          ...          ...   
2022-12-27 00:00:00    28.492051    28.771900    27.781102    27.957133   
2022-12-28 00:00:00    28.492051    28.771900    27.781102    27.957133   
2022-12-29 00:00:00    28.492051    28.771900    27.781102    27.957133   
2022-12-30 00:00:00    28.492051    28.771900    27.781102    27.957133   
2022-12-31 00:00:00    28.492051    28.771900    27.781102    27.957133   

STATION              IV000005557  IV000005562  ...  SGM00061679  SGM00061695  \
2014-05-26 00:00:00    26.923734    28.364358  ...    30.180399    27.984794   
2014-05-27 00:00:00    26.965717    27.881670  ...    29.645504    27.534573   
2014-05-28 00:00:00    27.106777    27.986913  ...    29.488717    27.975913   
2014-05-29 00:00:00    26.960080    28.104253  ...    29.663404    28.123871   
2014-05-30 00:00:00    26.935293    28.080711  ...    29.812440    28.240910   
...                          ...          ...  ...          ...          ...   
2022-12-27 00:00:00    26.128001    27.442904  ...    28.715670    27.013478   
2022-12-28 00:00:00    26.128001    27.442904  ...    28.715670    27.013478   
2022-12-29 00:00:00    26.128001    27.442904  ...    28.715670    27.013478   
2022-12-30 00:00:00    26.128001    27.442904  ...    28.715670    27.013478   
2022-12-31 00:00:00    26.128001    27.442904  ...    28.715670    27.013478   

STATION              SGM00061697  SGM00061698  SGM00061699  SL000061856  \
2014-05-26 00:00:00    25.263010    31.418281    31.622774    27.480532   
2014-05-27 00:00:00    25.234681    30.901097    30.879395    27.678423   
2014-05-28 00:00:00    25.309921    31.007351    31.008769    27.987301   
2014-05-29 00:00:00    25.506873    31.217753    31.131254    27.931188   
2014-05-30 00:00:00    25.725906    31.357104    31.260627    27.916198   
...                          ...          ...          ...          ...   
2022-12-27 00:00:00    26.259338    29.023683    29.351528    27.081878   
2022-12-28 00:00:00    26.259338    29.023683    29.351528    27.081878   
2022-12-29 00:00:00    26.259338    29.023683    29.351528    27.081878   
2022-12-30 00:00:00    26.259338    29.023683    29.351528    27.081878   
2022-12-31 00:00:00    26.259338    29.023683    29.351528    27.081878   

STATION              LIM00065660  GHM00065472  CVM00008594  GV000001832  
2014-05-26 00:00:00    27.645150    28.082508    23.349293    28.047105  
2014-05-27 00:00:00    27.617167    27.842467    23.259552    28.228056  
2014-05-28 00:00:00    27.550232    27.871235    23.440326    28.296795  
2014-05-29 00:00:00    27.645591    27.7

In [31]:
#Checking for the perfomance
import numpy as np

from sklearn.metrics import mean_squared_error, mean_absolute_error

# Define a function for RMSE
def rmse(test, pred_df):
    return np.sqrt(mean_squared_error(test, pred_df))

# Calculate and print error metrics for each time series
for column in test.columns:
    print(f'Error metrics for {column}:')
    mae = mean_absolute_error(test[column], pred_df[column])
    mse = mean_squared_error(test[column], pred_df[column])
    _rmse = rmse(test[column], pred_df[column])
    
    print(f'MAE: {mae}')
    print(f'MSE: {mse}')
    print(f'RMSE: {_rmse}\n')

Error metrics for UV000005502:
MAE: 2.8137195774507058
MSE: 11.801577170477996
RMSE: 3.4353423658316786

Error metrics for UV000005507:
MAE: 2.4322438081755835
MSE: 9.104068600750919
RMSE: 3.0172949144475285

Error metrics for UV000005522:
MAE: 1.9462635116689644
MSE: 5.932006100267156
RMSE: 2.4355710008675904

Error metrics for UV000065501:
MAE: 3.1226263929261515
MSE: 14.265654302899755
RMSE: 3.7769901115702904

Error metrics for UV000065516:
MAE: 2.3950981160000233
MSE: 8.797375916943189
RMSE: 2.9660370727526635

Error metrics for UVM00065503:
MAE: 2.524986384992721
MSE: 9.69633904545291
RMSE: 3.113894514182025

Error metrics for UVM00065510:
MAE: 2.0763376295510767
MSE: 6.470514753540052
RMSE: 2.543720651632182

Error metrics for UVM00065518:
MAE: 2.276313385948352
MSE: 8.089403491937176
RMSE: 2.844187668199336

Error metrics for IV000005557:
MAE: 1.225791132321324
MSE: 2.4432445735545483
RMSE: 1.5630881528418505

Error metrics for IV000005562:
MAE: 1.357725368126905
MSE: 2.9455847

A lower MAE indicates better accuracy. For example, for station IV000005557, the MAE is 1.2258, meaning that, on average, the model's predictions were off by about 1.2258 degrees.

This is similar to MAE but squares the differences before averaging them. This squaring gives more weight to larger errors. It's useful when you are more concerned about large errors. For instance, for station NG000001080, the MSE is 18.1342, indicating that when errors are squared, the average squared error is about 18.1342. Higher values indicate poorer model performance, especially in terms of larger errors.

This is the square root of the MSE. It brings the error metric back to the same scale as the data, making it more interpretable. RMSE is a good measure of the accuracy of the model and, like MSE, gives higher weight to larger errors. A lower RMSE value indicates better fit. For example, station NG000061017 has an RMSE of 7.0118, meaning the standard deviation of the prediction errors is about 7.0118 degrees.

# So in conclusion, the model is good and from the test done above, we can see that most of the station are only off by around 1 to 3 degrees. 